### PR TITLE
Adding an alias `pose proof (x:=t)` for `pose proof t as x` following the model of `pose (x:=t)`.

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

This was considered to be an intuitive syntax in a discussion at [HoTT/HoTT](https://github.com/HoTT/HoTT/pull/1208#discussion_r374342962).

- [X] Added / updated test-suite
- [X] Corresponding documentation was added / updated
- [X] Entry added in the changelog

### CODE REVIEW

**Comment by ppedrot** [comment]
I am not sure I understand the syntactic gain compared to `assert (x := t)`. The proposed syntax is longer, less mechanically parsable and furthermore conflicts with `pose`. If anything, I'd rather get rid of `pose proof`.

**Comment by herbelin** [comment]
> I am not sure I understand the syntactic gain compared to `assert (x := t)`. The proposed syntax is longer, less mechanically parsable and furthermore conflicts with `pose`. If anything, I'd rather get rid of `pose proof`.

The motivation is that some users find it intuitive and I agree with their arguments (an intuitive analogy `pose`/`pose proof`). I also agree that ihis is not a necessary syntax per se.

Actually, I'd be happy that you give your view about tactics in general. There is room for making the set of primitive tactics much better (better organized, more powerful) and that would be good that we adopt a coherent vision of where we are going. In particular, if you could give your views, that would help to understand what motivates your position on this PR.

**Review by Zimmi48 — APPROVED** [review]
I welcome this change. The inconsistency that there is today between the syntax of `pose` and `pose proof` is not justified and the duality `pose` / `pose proof` makes sense. Having synonyms is not a big deal *per se*. What is problematic is when there are subtle and unclear differences between many variants.

**Comment by Zimmi48** [comment]
> I'd rather get rid of `pose proof`.

This is the other option, but in this case, we should likely get rid not just of `pose proof` but `pose` as well (because there is already `set`). As long as we decide not to remove `pose proof`, we should make it more consistent with `pose`.

**Comment by SkySkimmer** [comment]
`pose` and `set` are not the same thing though?

**Comment by ppedrot** [comment]
`set` is in theory stronger, as you can write `set ... in |-` for `pose`. But the syntax is cumbersome.

**Comment by Zimmi48** [comment]
Alright, I was confused. Let's forget about removing `pose` and just settle on whether we want to keep `pose proof` and merge this PR, or formally deprecate `pose proof`.

**Comment by ppedrot** [comment]
I don't mind merging this PR locally for the sake of uniformity, but I think that eventually these redundant syntaxes should be deprecated.

**Comment by herbelin** [comment]
> but I think that eventually these redundant syntaxes should be deprecated.

No opposition to that, to the condition that we have a nice picture to propose. I'm confident that, the day we are ready for, we can provide a nice picture about syntax (and semantics) of tactics, leading in particular to keep the "good" syntax among the set of redundant ones. My current question is: how to get closer to that day? @ppedrot, you obviously have a view in mind regarding tactics, but I have some difficulties to decipher it without your help.

**Comment by ejgallego** [comment]
Is `pose proof` exactly the same than `assert` ?

**Comment by herbelin** [comment]
> Is `pose proof` exactly the same than `assert` ?

`pose proof p as x` is exactly the same than `assert (x:=p)` but `pose proof p as pat`, for `pat` an introduction pattern, has no equivalent using `assert`.

The proposed `pose proof (x:=p)` is exactly the same as `assert (x:=p)`. The reason for a different syntax is for the analogy with `pose (x:=c)` which was a concern in the HoTT discussion mentioned above. Note that `pose c as x` is itself already a synonym of `pose (x:=c)`, which is another argument to expect the same with `pose proof`.

I'm looking for hearing about how the use of `as` is perceived. Somehow, the fact that `(id:=c)` does not scale nicely to `(pat:=c)` and that this leads to two syntaxes `(x:=c)` and `c as x` is a problem. For instance, `pose proof (<- := c)` / `assert (<- := c)` does not look so good, esthetically speaking. In comparison, the absence of parentheses in `have` makes it a bit more acceptable, as in `have <- := c`, so dropping parentheses in `assert` could somehow be a compromise in the long term (up to the loss of unity of style).

**Comment by ejgallego** [comment]
Indeed having then both `pose proof` and `assert` seems a bit redundant then; IMHO having a `pat` on the left side at first looks strange but later on is very convenient and a bit more natural in the flow than `as`.

Example `pose proof [a,b] := foo.` or `assert [a,b] := foo.`

Dropping the parenthesis seems doable as ssreflect has shown so why not? That could be a nice step towards convergence.

**Comment by herbelin** [comment]
> Indeed having then both `pose proof` and `assert` seems a bit redundant then

It seems redundant but it is not clear if we replace them in context. First, they have a different intuitive intent, as in `Axiom`/`Theorem`. Second, they are technically not mergeable in their simplest versions `pose proof p` and `assert T` since there is then no `:` or `:=` to distinguish them. More generally, what is the question?

@ppedrot said: "I think that eventually these redundant syntaxes should be deprecated." but I fail to understand what it means concretely. `Lemma` and `Theorem` are redundant, and `Definition` and `Theorem` are redundant also to some respect. What is the problem with `pose proof` and `assert`? Is the problem a problem of redundancy, a problem of unity, a problem of clarity, a problem of grammar, a problem of how to teach them?

I'm supporting the PR because it replaces/complements in an intuitive way the hole currently filled by **assert (x:=p)** in the following  picture

assert T            |   assert (x:T)             |    assert T as pat |
:----|:----|:----
**pose proof p**   |   **?**   |   **pose proof p as pat** |

But we can actually generalize and integrate **pose** to the picture and this gives:

assert T            |   assert (x:T)             |    assert T as pat | |
:----|:----|:----|:----
**pose proof p**   |   **?**   |   **pose proof p as pat** | 
**pose t** | **pose (x:=t)**  | **pose t as pat** | **pose (f binders := t)** 

with the restriction that **pat** in **pose t as pat** has to be an identifier (this is maybe excessive: we virtually might want to do **pose (f 0) as (x,y)** for **f:nat -> nat*nat**).

This suggests that, for consistency, I could also add:
- **pose t as pat** (any **pat**)
- **assert (f binders : T)** (which ssreflect **have:** provides actually)
- **pose proof (f binders := t)** (which ssreflect **have:=** provides)

@ppedrot: would you like it better if the PR would add also the three above extensions, so as to complete the picture further?

**Comment by herbelin** [comment]
> That could be a nice step towards convergence.

About a convergence with ssreflect, the ssreflect story is:

have: T            |   have x : T             |    have pat : T |  have f binders : T
:----|:----|:----|:----
**have:= p**   |   **have x := p**   |   **have pat := p** | **have f binders := p**
**have: T := p**   |   **have x : T := p**   |   **have pat : T := p** | **have f binders : T := p**
| | **pose x := t** | | **pose f binders := t**
| | **pose x : T := t** | | **pose f binders : T := t**

which is very nice in many respects.

(_I made simplifications which I think are secondary: **pat** is morally one pattern but ssreflect parser takes a list of patterns; the syntax of **binders** is not the standard one but an ad hoc one._)

The obvious basis to satisfy the legitimate user expectation for a unified system of tactics would be to alias **assert** and **have:**, and to alias **pose proof** and **have:=**. Then:
- There is a conflict between dropping parentheses and allowing the beginner-friendly **assert T** and **pose proof T**. Telling the beginners that they have to use **assert: T** or **assert x : T** or **assert ? : T** is maybe not too much asking, I don't know.
- The presence of parentheses is indeed also a visual obstacle to the use of **pat** in first position.

> Dropping the parenthesis seems doable as ssreflect has shown so why not? That could be a nice step towards convergence.

So maybe, the reasonable direction is to drop parentheses, renouncing to the `pose proof p` and `assert T` form, even though they are "friendly", to the benefit of a lighter syntax, making the correspondence with `Theorem f binders : T` and `Definition f binders := t` closer, allowing the `f binders` subexpression to be replaced by a `pat`.

In principle, `pose proof p` and `assert T` could actually be kept since `p` and `T` are at level 8 and that this does not conflict with the other syntaxes per se, but that would make a pretty tricky grammar. If they are deprecated, then, the question of `pose proof` vs `assert` comes back, and, obviously, the question is not only technical but also about what the words convey.

**Comment by ejgallego** [comment]
Thanks for the detailed overview @herbelin , indeed I'm a big fan of `have`, some of the reasons for me are:
- it is very short to type
- the "cognitive load" is pretty low; whenever I want to put something in the context (*) my brain thinks of both forms of "have" as a single entity, the `:` vs `:=` distinction is akin to giving a body to a definition or going to proof mode which I find natural. Somehow the left hand side is handled at my cognitive level as a single "pattern" hole

At least in my case more syntactic uniformity makes the tactic easier to use for me.

(*) then of course there is `set` which I memo as "alias stuff" , as oppose to "declare stuff"

**Comment by herbelin** [comment]
> indeed I'm a big fan of have, some of the reasons for me are

What to do then? A large difficult-to-evaluate number of users are using `assert` and another large difficult-to-evaluate number of users are using `have`. What can we concretely propose to them for a convergence?

I mean, what can we concretely propose to them other than first joining the semantics, unifying the syntax, and, eventually, telling, oh, these are synonyms, you can pass a `sed` on your files to keep only one name?

**Comment by ejgallego** [comment]
> I mean, what can we concretely propose to them other than first joining the semantics, unifying the syntax, and, eventually, telling, oh, these are synonyms, you can pass a `sed` on your files to keep only one name?

Actually that seems like a pretty good plan!

Programming style questions are in a large part a matter of taste tho.

**Comment by ejgallego** [comment]
> Actually that seems like a pretty good plan!

In the short term we should start with having `pose / assert` to be aliases.

This may create problem so indeed maybe it is the time we require some kind of header "tactic-version: 1" for `.v` files.

Haskell does a pretty good job with compiler pragmas in the headers, although not perfect for example due to the large amount of options sometimes users need to declare.

**Comment by Zimmi48** [comment]
Haskell pragmas are just like our own flags and options except they must be in the header right? I don't know why we would like to make our system weaker. Locally applying options is great for incremental conversion of a file to a new syntax / language / heuristics / whatever.

**Comment by ejgallego** [comment]
They can be more powerful as they can be processed before the actual document setup takes place.

**Comment by herbelin** [comment]
> Actually that seems like a pretty good plan!

Thanks for your support! For making the semantics closer, I can see already two issues:
  - the different semantics of intro-patterns;
  - the automatic generalization made over free evars in the rhs of `have` and `pose`.

Regarding the semantics of intro-patterns, there are several solutions. 
  - first, my guess is that most of the time, ssreflect users gives complete patterns, so that, in practice vanilla auto-completion according to the structure of the type would work the same;
  - we can also have an option to turn on/off structural completion;
  - another alternative would be to use the keyword `have` vs `assert` to tell the intended semantic of the introduction pattern.

Regarding the automatic generalization made over free evars, this is a rather intuitive feature. A canonical question is whether we want it on by default or not. Somehow, the `` ` `` has already this meaning of generalizing, so this can be a direction to investigate (or using another convention to deactivate it).

Otherwise there is the question of binders, and it seems natural that in all cases we accept all kinds of binders (including e.g. `'(x,y)`).

What other differences are there between `have` and `assert`?

Incidentally, in which sense this PR is not a step towards merging the different semantics and expressiveness of the syntax we have?

**Comment by Zimmi48** [comment]
>another alternative would be to use the keyword `have` vs `assert` to tell the intended semantic of the introduction pattern

Maybe in the beginning, but in the longer term it would be better not to have to remember this and to instead have this controlled by a flag or to remove this difference entirely.

**Comment by ppedrot** [comment]
Anyways, it seems that this should be merged. I am assigning and will merge soon.

**Review by ppedrot — APPROVED** [review]

File: doc/changelog/04-tactics/11522-master+pose-proof-wo-as-syntax.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+- **Added:**
+  Syntax :tacn:`pose proof (@ident:=@term)` as an
```
Comment (Zimmi48) [inline doc/changelog/04-tactics/11522-master+pose-proof-wo-as-syntax.rst]: Here you actually meant to use the `:n:` (for notation) quoting context:

```suggestion
  Syntax :n:`pose proof (@ident:=@term)` as an
```

File: doc/changelog/04-tactics/11522-master+pose-proof-wo-as-syntax.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+  alternative to `pose proof @term as @ident`, following the model of
```
Comment (Zimmi48) [inline doc/changelog/04-tactics/11522-master+pose-proof-wo-as-syntax.rst]:
```suggestion
  alternative to :n:`pose proof @term as @ident`, following the model of
```

File: doc/changelog/04-tactics/11522-master+pose-proof-wo-as-syntax.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+  :tacn:`pose (@ident:=@term)`
```
Comment (Zimmi48) [inline doc/changelog/04-tactics/11522-master+pose-proof-wo-as-syntax.rst]: For the link to the doc, let's still include a `tacn`.
```suggestion
  :n:`pose (@ident:=@term)`. See documentation of :tacn:`pose proof`
```

**Pull request merged**